internal/dag/secret.go

```diff
@@ -50,35 +51,47 @@ func isValidSecret(secret *v1.Secret) (bool, error) {
 }
 
 func validateCertificate(data []byte) error {
-	cert, err := decodePEM(data)
-	if err != nil {
-		return err
-	}
-	_, err = x509.ParseCertificate(cert.Bytes)
-	return err
-}
-
-func validatePrivateKey(data []byte) error {
-	key, err := decodePEM(data)
-	if err != nil {
-		return nil
-	}
-	if _, err := x509.ParsePKCS1PrivateKey(key.Bytes); err == nil {
-		return nil
-	}
-	if _, err := x509.ParsePKCS8PrivateKey(key.Bytes); err == nil {
-		return nil
+	var exists bool
+	var block *pem.Block
+	for rest := data[:]; len(rest) > 0; {
+		block, rest = pem.Decode(rest)
+		if block == nil {
+			return errors.New("failed to parse PEM block")
+		}
+		if !strings.HasSuffix(block.Type, "CERTIFICATE") {
+			continue
+		}
+		if _, err := x509.ParseCertificate(block.Bytes); err != nil {
+			return err
+		}
+		exists = true
 	}
-	if _, err := x509.ParseECPrivateKey(key.Bytes); err == nil {
-		return nil
+	if !exists {
+		return errors.New("failed to locate certificate")
 	}
-	return errors.New("unknown private key encoding")
+	return nil
 }
 
-func decodePEM(in []byte) (*pem.Block, error) {
-	block, _ := pem.Decode(in)
-	if block == nil {
-		return nil, errors.New("failed to parse PEM")
+func validatePrivateKey(data []byte) error {
+	var block *pem.Block
+	for len(data) > 0 {
+		block, data = pem.Decode(data)
+		if block == nil {
+			return errors.New("failed to parse PEM block")
+		}
+		if !strings.HasSuffix(block.Type, "PRIVATE KEY") {
+			continue
```

Comment:
if the private key doesn't contain a private key, should that be an error?

Reply:
Looking at the example you provided I think it would be better if rather than ignoring blocks that didn't start with `PRIVATE KEY` the logic was inverted and it excluded only blocks it was expecting, ie, `EC PARAMETERS`

Reply:
If the private key doesn't contain a private key, the loop will terminate on `len(data)` and return `errors.New("failed to locate private key")`.  

The way I've structured it is likely leading to confusion.  Currently, it is seek to first block that ends with `PRIVATE KEY` (covering `PRIVATE KEY`, `RSA PRIVATE KEY` and `EC PRIVATE KEY`). Validate the block bytes as a key and return either `nil` or `errors.New("unknown private key encoding")`.

I'm a little worried about inversion (though I'm not against it). I didn't want to miss a block type or have a new block type show up that is treated/validated/rejected as a key.

Reply:
Would something like this work?
```
func validatePrivateKey(data []byte) error {
        var keys int
        for len(data) > 0 {
                var block *pem.Block
                block, data = pem.Decode(data)
                if block == nil {
                        return errors.New("failed to parse PEM block")
                }
                switch block.Type {
                case "PRIVATE KEY":
                        if _, err := x509.ParsePKCS8PrivateKey(block.Bytes); err != nil {
                                return err
                        }
                        keys++
                case "RSA PRIVATE KEY":
                        if _, err := x509.ParsePKCS1PrivateKey(block.Bytes); err != nil {
                                return err
                        }
                        keys++
                case "EC PRIVATE KEY":
                        if _, err := x509.ParseECPrivateKey(block.Bytes); err != nil {
                                return err
                        }
                        keys++
                case "EC PARAMETERS":
                        // ignored
                default:
                        return errors.New("unknown private key type: " + block.Type)
                }
        }
        switch keys {
        case 0:
                return errors.New("failed to locate private key")
        case 1:
                return nil
        default:
                return errors.New("multiple private keys")
        }
}
```

Reply:
I was trying to avoid those string compares :P . Feels like you want to add
```
type BlockType string
const (
	PrivateKey BlockType = "PRIVATE KEY"
	RSAPrivateKey BlockType = "RSA PRIVATE KEY"
	ECPrivateKey BlockType = "EC PRIVATE KEY"
	Certificate BlockType = "CERTIFICATE"
)

switch BlockType(block.Type) {
}
```

From what I can tell its close.  It appears as though, it is common to attempt parsing "PRIVATE KEY" and "RSA PRIVATE KEY" with both `ParsePKCS1PrivateKey` and `ParsePKCS8PrivateKey`

Reply:
What do you think about this?

```
func validatePrivateKey(data []byte) error {
	var keys int
	for len(data) > 0 {
		var block *pem.Block
		block, data = pem.Decode(data)
		if block == nil {
			return errors.New("failed to parse PEM block")
		}
		switch block.Type {
		case "PRIVATE KEY":
			if err := validatePKCSPrivateKey(block.Bytes); err != nil {
				return err
			}
			keys++
		case "RSA PRIVATE KEY":
			if err := validatePKCSPrivateKey(block.Bytes); err != nil {
				return err
			}
			keys++
		case "EC PRIVATE KEY":
			if _, err := x509.ParseECPrivateKey(block.Bytes); err != nil {
				return err
			}
			keys++
		}
	}
	switch keys {
	case 0:
		return errors.New("failed to locate private key")
	case 1:
		return nil
	default:
		return errors.New("multiple private keys")
	}
}

func validatePKCSPrivateKey(in []byte) error {
	if _, err := x509.ParsePKCS1PrivateKey(in); err == nil {
		return nil
	}
	if _, err := x509.ParsePKCS8PrivateKey(in); err == nil {
		return nil
	}
	return errors.New("unknown PKCS encoding")
}
```
 > subtly different, if it is not a block it knows how to parse/validate, it skips it.